### PR TITLE
Stricter mypy checking

### LINKED ISSUE #331: Enable strict mypy checking
A follow-on from #326. Using the `--strict` flag with `mypy` produces another 98 errors that will need to be squashed.

### PR DESCRIPTION
**Pull request by karalekas**
Description
-----------

Fixes #331 in line with the discussion in that issue.

Checklist
-----------
Please make sure you have finished the following tasks before requesting a review of the pull request (PR). For more information, please refer to the [Contributor's Guide](../blob/master/CONTRIBUTING.md):

- [x] Contributions to mitiq should follow the [pep8 style](https://www.python.org/dev/peps/pep-0008/). You can enforce it easily with [`black`](https://black.readthedocs.io/en/stable/index.html) style and with [`flake8`](http://flake8.pycqa.org) conventions.
- [x] Please add tests to cover your changes, if applicable, and make sure that all new and existing tests pass locally.
- [x] Functions and classes have useful [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings and [type hints](https://www.python.org/dev/peps/pep-0484/) in the signature of the objects.
- [x] The [changelog](https://github.com/unitaryfund/mitiq/blob/master/CHANGELOG.md) is updated, including author and PR number (@username, gh-xxx).

### CODE REVIEW

**Comment by karalekas** [comment]
I think that we should ignore codecov here.

**Comment by karalekas** [comment]
Will likely need to fix some merge conflicts after #381 is merged.

File: mitiq/zne/inference.py [hunk 1]
```diff
@@ -302,7 +304,9 @@ def _noise_to_expval(scale_factor, **exec_params) -> float:
 
         return self.iterate(_noise_to_expval, max_iterations)
 
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Factory):
+            return False
```
Comment (willzeng) [inline mitiq/zne/inference.py:309]: Does it make sense to allow comparisons to things that aren't Factory type objects? Should it be
```suggestion
    def __eq__(self, other: Factory) -> bool:
        if not isinstance(other, Factory):
            return False
```

Comment (willzeng) [reply mitiq/zne/inference.py:309]: This applies to a few other __eq__ methods below but I won't duplicate the comment

**Review by willzeng — APPROVED** [review]
One small question on this that is not blocking. I also agree that we can ignore that the coverage slightly goes down here because of some branching on type checks.

Nicely done! Great to have this as part of our CI @karalekas

File: mitiq/zne/inference.py (hunk 1 above)
Comment (karalekas) [reply mitiq/zne/inference.py:309]: That's a great question! It should in fact be `Any`, and if you change it to something else here's what happens:
```
mitiq/zne/inference.py:757: error: Argument 1 of "__eq__" is incompatible with supertype "object"; supertype defines the argument type as "object"
mitiq/zne/inference.py:757: note: This violates the Liskov substitution principle
mitiq/zne/inference.py:757: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
mitiq/zne/inference.py:757: note: It is recommended for "__eq__" to work with arbitrary objects, for example:
mitiq/zne/inference.py:757: note:     def __eq__(self, other: object) -> bool:
mitiq/zne/inference.py:757: note:         if not isinstance(other, ExpFactory):
mitiq/zne/inference.py:757: note:             return NotImplemented
mitiq/zne/inference.py:757: note:         return <logic to compare two ExpFactory instances>
```
The equality comparison method is defined by `object` to accept any type as `other` and thus it violates the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) to change it (this is one of the tenets of good object-oriented SWE, part of a collection called [SOLID](https://en.wikipedia.org/wiki/SOLID)).

**Comment by willzeng** [comment]
Sounds good

**Comment by nathanshammah** [comment]
This SOLID tenets, is it something we’d like to add to guidelines for contributions? Looks interesting and I was not aware of them, thanks @karalekas . https://github.com/unitaryfund/mitiq/pull/380#discussion_r498998800

**Pull request merged**